### PR TITLE
Assignment9 mmwoodfo & Veltoc

### PR DESCRIPTION
**Pull request by mmwoodfo**
With @Veltoc

### CODE REVIEW

File: SecretSanta/test/SecretSanta.Web.Tests/GiftTests.cs [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+            ApiHostProcess.WaitForExit(8000);
+
+            //AddUser
+            HttpClient httpClient = new HttpClient();
+            httpClient.BaseAddress = new Uri("http://localhost:5000/");
+            httpClient.DefaultRequestHeaders.Accept.Clear();
+            httpClient.DefaultRequestHeaders.Accept.Add(
+                new MediaTypeWithQualityHeaderValue("application/json"));
+
+            UserClient userClient = new UserClient(httpClient);
```
Comment (mmwoodfo) [inline SecretSanta/test/SecretSanta.Web.Tests/GiftTests.cs]: 1. Add a user via the API (this is the only step not using Selenium) ✔

File: SecretSanta/test/SecretSanta.Web.Tests/GiftTests.cs [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+        public void TakeScreenShot(string fileName)
+        {
+            ((ITakesScreenshot)Driver).GetScreenshot().SaveAsFile($"{fileName}.png", ScreenshotImageFormat.Png);
+        }
+
+        //----------------- CREATE GIFTS -----------------//
+
+        public void ClickCreateButton()
+        {
+            Driver.Navigate().GoToUrl(new Uri(AppUrl));
```
Comment (mmwoodfo) [inline SecretSanta/test/SecretSanta.Web.Tests/GiftTests.cs]: 2. Navigate to the ~/Gifts Page. ✔

File: SecretSanta/test/SecretSanta.Web.Tests/GiftTests.cs [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+            IWebElement element = Driver.FindElement(By.Id("create-gift-btn"));
+            element.Click();
```
Comment (mmwoodfo) [inline SecretSanta/test/SecretSanta.Web.Tests/GiftTests.cs]: 3. Find the Create Gift button **using a CSS Selector** and click it. ✔

File: SecretSanta/test/SecretSanta.Web.Tests/GiftTests.cs [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        }
+
+        public void EnterGiftInformation(string title, string description, string url, int index)
+        {
+            IWebElement giftTitleElement = Driver.FindElement(By.Id("gift-title"));
+            IWebElement giftDescriptionElement = Driver.FindElement(By.Id("gift-description"));
+            IWebElement giftUrlElement = Driver.FindElement(By.Id("gift-url"));
+            IWebElement giftUserSelect = Driver.FindElement(By.Id("gift-user-dropdown"));
+
+            //Add gift information
```
Comment (mmwoodfo) [inline SecretSanta/test/SecretSanta.Web.Tests/GiftTests.cs]: 4. Enter the gift information. ✔

File: SecretSanta/test/SecretSanta.Web.Tests/GiftTests.cs [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+            //Assert
+            Assert.AreEqual<string>(title, giftTitleElement.GetProperty("value"));
+            Assert.AreEqual<string>(description, giftDescriptionElement.GetProperty("value"));
+            Assert.AreEqual<string>(url, giftUrlElement.GetProperty("value"));
+        }
+
+        public void ClickSubmitButton()
+        {
+            IWebElement element = Driver.FindElement(By.Id("submit"));
+            element.Click();
```
Comment (mmwoodfo) [inline SecretSanta/test/SecretSanta.Web.Tests/GiftTests.cs]: 5. Find the Submit button **via Id** and Submit. ✔

File: SecretSanta/test/SecretSanta.Web.Tests/GiftTests.cs [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+            ClickCreateButton();
+            EnterGiftInformation(title, description, url, 0);
+            ClickSubmitButton();
+            Thread.Sleep(5000);
+
+            //Act
+            var giftAttributes = GetListOfGifts();
+            int index = giftAttributes.IndexOf(title);
+
+            //Assert
```
Comment (mmwoodfo) [inline SecretSanta/test/SecretSanta.Web.Tests/GiftTests.cs]: 6. Verify the gift appears in the list. ✔

File: SecretSanta/test/SecretSanta.Web.Tests/GiftTests.cs [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+            Assert.AreEqual(title, giftAttributes[index]);
+            Assert.AreEqual(description, giftAttributes[index + 1]);
+            Assert.AreEqual(url, giftAttributes[index + 2]);
+
+            //Take screen shot upon success
```
Comment (mmwoodfo) [inline SecretSanta/test/SecretSanta.Web.Tests/GiftTests.cs]: 7. Take a screen shot of the updated Gifts list (using Selenium). ✔

File: azure-pipelines.yml [hunk 8]
```diff
@@ -1,19 +1,18 @@
-# ASP.NET Core
-# Build and test ASP.NET Core projects targeting .NET Core.
-# Add steps that run tests, create a NuGet package, deploy, and more:
-# https://docs.microsoft.com/azure/devops/pipelines/languages/dotnet-core
-
 trigger:
```
Comment (mmwoodfo) [inline azure-pipelines.yml]: 1. Update the Azure DevOps pipeline to run your selenium tests

File: SecretSanta/src/SecretSanta.Web/SecretSanta.Web.csproj [hunk 9]
```diff
@@ -21,4 +21,13 @@
     <TypeScriptCompile Include="scripts\vue-shim.d.ts" />
   </ItemGroup>
 
+  <ItemGroup>
+    <Folder Include="wwwroot\css\" />
+    <Folder Include="wwwroot\js\" />
+  </ItemGroup>
+  
+<Target Condition="!Exists('Views\Shared\_Layout.cshtml')" Name="PostBuild" AfterTargets="PostBuildEvent">
+    <Exec WorkingDirectory="$(MSBuildProjectDirectory)" Command="npm run build:prod" />
```
Comment (mmwoodfo) [inline SecretSanta/src/SecretSanta.Web/SecretSanta.Web.csproj]: 2. Consider adding `npm run build:prod` to your csproj file as a post build step. ✔

File: SecretSanta/test/SecretSanta.Web.Tests/GiftTests.cs [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+            string path = $"{Directory.GetCurrentDirectory()}CreateGiftTest.png";
+            ((ITakesScreenshot)Driver).GetScreenshot().SaveAsFile(path, ScreenshotImageFormat.Png);
+            this.TestContext.AddResultFile(path);
+        }
+
+        ////----------------- VALIDATE LINKS -----------------//
+        [TestMethod]
+        [TestCategory("Chrome")]
+        public void ValidateLinks_GiftsListPage()
```
Comment (mmwoodfo) [inline SecretSanta/test/SecretSanta.Web.Tests/GiftTests.cs]: - Add a test that verifies all links within the app are valid. ✔

Works locally, but doesn't work on Azure. Trying to figure out why still. Upon investigation, the tests on Azure grab the links correctly (starting with the headers Secret Santa home link), but the test is unable to verify the http link as being valid or not.

File: SecretSanta/test/SecretSanta.Web.Tests/GiftTests.cs [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+
+            //Assert
+            Assert.AreEqual(title, giftAttributes[index]);
+            Assert.AreEqual(description, giftAttributes[index + 1]);
+            Assert.AreEqual(url, giftAttributes[index + 2]);
+
+            //Take screen shot upon success
+            string path = $"{Directory.GetCurrentDirectory()}CreateGiftTest.png";
+            ((ITakesScreenshot)Driver).GetScreenshot().SaveAsFile(path, ScreenshotImageFormat.Png);
+            this.TestContext.AddResultFile(path);
```
Comment (mmwoodfo) [inline SecretSanta/test/SecretSanta.Web.Tests/GiftTests.cs]: 3. Manually create a screen shot of **Azure DevOps** displaying the Gift list screen shot with the added gift in your PR. ✔

File: SecretSanta/test/SecretSanta.Web.Tests/GroupTests.cs [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+        [TestInitialize]
+        public void TestInitialize()
+        {
+            Driver = new ChromeDriver();
+            Driver.Manage().Timeouts().ImplicitWait = new System.TimeSpan(0, 0, 10);
+        }
+
+        [TestMethod]
+        [TestCategory("Chrome")]
+        public void ValidateLinks_GiftsListPage()
```
Comment (Veltoc) [inline SecretSanta/test/SecretSanta.Web.Tests/GroupTests.cs]: Same test names as gifts. same for Users and Home

File: SecretSanta/test/SecretSanta.Web.Tests/GiftTests.cs [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+        private static Process? WebHostProcess { get; set; }
+        string AppUrl { get; } = "http://localhost:5001/Gifts";
+
+        [ClassInitialize]
+        public static async Task ClassInitalize(TestContext testContext)
+        {
+            if (testContext is null)
+                throw new ArgumentNullException(nameof(testContext));
+
+            ApiHostProcess = Process.Start("dotnet.exe", "run -p ..\\..\\..\\..\\..\\src\\SecretSanta.Api\\SecretSanta.Api.csproj");
```
Comment (twofingerrightclick) [inline SecretSanta/test/SecretSanta.Web.Tests/GiftTests.cs]: drop the exe for the ubuntu vm thats on azure

**Pull request closed without merging**